Misleading error message when attempting to read binary VTK files

Meshio seems to only support ascii VTK files, not binary. This would not be a problem, but the error that is printed does not give any clue that the issue is that the file is binary.

Example:

```
>>> mesh = meshio.read(r"c:\tmp\model-ascii.vtk")
>>> mesh = meshio.read(r"c:\tmp\model-bin.vtk")
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "C:\Users\andra\AppData\Local\NA-MIC\Slicer 4.11.0-2019-07-30\lib\Python\Lib\site-packages\meshio\_helpers.py", line 174, in read
    return format_to_reader[file_format].read(filename)
  File "C:\Users\andra\AppData\Local\NA-MIC\Slicer 4.11.0-2019-07-30\lib\Python\Lib\site-packages\meshio\_vtk.py", line 172, in read
    out = read_buffer(f)
  File "C:\Users\andra\AppData\Local\NA-MIC\Slicer 4.11.0-2019-07-30\lib\Python\Lib\site-packages\meshio\_vtk.py", line 206, in read_buffer
    _read_sub_section(f, info)
  File "C:\Users\andra\AppData\Local\NA-MIC\Slicer 4.11.0-2019-07-30\lib\Python\Lib\site-packages\meshio\_vtk.py", line 295, in _read_sub_section
    assert info.section == "FIELD", "Unknown section '{}'.".format(info.section)
AssertionError: Unknown section '
```

Example ascii and binary files are available here: https://1drv.ms/f/s!Arm_AFxB9yqHt_BH5m1ZSSfibgVrOg

--- a/meshio/__about__.py
+++ b/meshio/__about__.py
@@ -1,4 +1,4 @@
-__version__ = "3.0.5"
+__version__ = "3.0.6"
 __author__ = "Nico Schlömer"
 __author_email__ = "nico.schloemer@gmail.com"
 __copyright__ = "Copyright (c) 2015-2019 {} <{}>".format(__author__, __author_email__)

--- a/meshio/_vtk.py
+++ b/meshio/_vtk.py
@@ -203,7 +203,7 @@ def read_buffer(f):
         if info.section in vtk_sections:
             _read_section(f, info)
         else:
-            _read_sub_section(f, info)
+            _read_subsection(f, info)
 
     _check_mesh(info)
     cells, cell_data = translate_cells(info.c, info.ct, info.cell_data_raw)
@@ -260,7 +260,7 @@ def _read_section(f, info):
         rgba = data.reshape((info.num_items, 4))  # noqa F841
 
 
-def _read_sub_section(f, info):
+def _read_subsection(f, info):
     if info.active == "POINT_DATA":
         d = info.point_data
     elif info.active == "CELL_DATA":
@@ -286,11 +286,11 @@ def _read_sub_section(f, info):
                 info.section, len(d[info.section])
             )
     elif info.section == "SCALARS":
-        d.update(_read_scalar_field(f, info.num_items, info.split))
+        d.update(_read_scalar_field(f, info.num_items, info.split, info.is_ascii))
     elif info.section == "VECTORS":
-        d.update(_read_field(f, info.num_items, info.split, [3]))
+        d.update(_read_field(f, info.num_items, info.split, [3], info.is_ascii))
     elif info.section == "TENSORS":
-        d.update(_read_field(f, info.num_items, info.split, [3, 3]))
+        d.update(_read_field(f, info.num_items, info.split, [3, 3], info.is_ascii))
     else:
         assert info.section == "FIELD", "Unknown section '{}'.".format(info.section)
         d.update(_read_fields(f, int(info.split[2]), info.is_ascii))
@@ -436,7 +436,7 @@ def _read_cell_types(f, is_ascii, num_items):
     return ct
 
 
-def _read_scalar_field(f, num_data, split):
+def _read_scalar_field(f, num_data, split, is_ascii):
     data_name = split[1]
     data_type = split[2].lower()
     try:
@@ -451,22 +451,39 @@ def _read_scalar_field(f, num_data, split):
     dtype = numpy.dtype(vtk_to_numpy_dtype_name[data_type])
     lt, _ = f.readline().decode("utf-8").split()
     assert lt.upper() == "LOOKUP_TABLE"
-    data = numpy.fromfile(f, count=num_data, sep=" ", dtype=dtype)
+
+    if is_ascii:
+        data = numpy.fromfile(f, count=num_data, sep=" ", dtype=dtype)
+    else:
+        # Binary data is big endian, see
+        # <https://www.vtk.org/Wiki/VTK/Writing_VTK_files_using_python#.22legacy.22>.
+        dtype = dtype.newbyteorder(">")
+        data = numpy.fromfile(f, count=num_data, dtype=dtype)
+        line = f.readline().decode("utf-8")
+        assert line == "\n"
 
     return {data_name: data}
 
 
-def _read_field(f, num_data, split, shape):
+def _read_field(f, num_data, split, shape, is_ascii):
     data_name = split[1]
     data_type = split[2].lower()
 
     dtype = numpy.dtype(vtk_to_numpy_dtype_name[data_type])
     # <https://stackoverflow.com/q/2104782/353337>
     k = reduce((lambda x, y: x * y), shape)
-    data = numpy.fromfile(f, count=k * num_data, sep=" ", dtype=dtype).reshape(
-        -1, *shape
-    )
 
+    if is_ascii:
+        data = numpy.fromfile(f, count=k * num_data, sep=" ", dtype=dtype)
+    else:
+        # Binary data is big endian, see
+        # <https://www.vtk.org/Wiki/VTK/Writing_VTK_files_using_python#.22legacy.22>.
+        dtype = dtype.newbyteorder(">")
+        data = numpy.fromfile(f, count=k * num_data, dtype=dtype)
+        line = f.readline().decode("utf-8")
+        assert line == "\n"
+
+    data = data.reshape(-1, *shape)
     return {data_name: data}
 
 

--- a/meshio/_vtu.py
+++ b/meshio/_vtu.py
@@ -380,13 +380,13 @@ def write(filename, mesh, write_binary=True, pretty_xml=True):
     # Don't use byteswap to make sure that the dtype is changed; see
     # <https://github.com/numpy/numpy/issues/10372>.
     points = mesh.points.astype(mesh.points.dtype.newbyteorder("="))
-    for data in mesh.point_data.values():
-        data = data.astype(data.dtype.newbyteorder("="))
+    for key, data in mesh.point_data.items():
+        mesh.point_data[key] = data.astype(data.dtype.newbyteorder("="))
     for data in mesh.cell_data.values():
-        for dat in data.values():
-            dat = dat.astype(dat.dtype.newbyteorder("="))
-    for data in mesh.field_data.values():
-        data = data.astype(data.dtype.newbyteorder("="))
+        for key, dat in data.items():
+            data[key] = dat.astype(dat.dtype.newbyteorder("="))
+    for key, data in mesh.field_data.items():
+        mesh.field_data[key] = data.astype(data.dtype.newbyteorder("="))
 
     def numpy_to_xml_array(parent, name, fmt, data):
         da = ET.SubElement(

